### PR TITLE
:sparkles: Update to Kubernetes 1.30

### LINKED ISSUE #3068: epic: Kubernetes 1.30
### Demo Objective

Opening this to track upgrading kcp and the related forks we currently maintain to Kubernetes 1.30, released in December 2023.

A similar epic for 1.28 was #3003.

### Stories

- [x] Update apimachinery
- [x] Update code-generator
- [x] Update client-go
- [x] update kubernetes
- [ ] update kcp

```[tasklist]
### Pull Requests
- [ ] https://github.com/kcp-dev/apimachinery/pull/79
- [ ] https://github.com/kcp-dev/code-generator/pull/90
- [ ] https://github.com/kcp-dev/client-go/pull/37
```

### PR DESCRIPTION
**Pull request by embik**
## Summary

This updates kcp to Kubernetes 1.30.0. The corresponding branch for our Kubernetes fork is [kcp-1.30](https://github.com/kcp-dev/kubernetes/tree/kcp-1.30). Mainly this PR adds new APIs or removes feature gates where they have been removed by upstream.

The bigger changes in this PR are:

- Updating to the new ValidatingAdmissionPolicy plugin infrastructure.
- Rewrite the OpenAPI controller we recently added to accommodate for new upstream types.

The following bigger TODOs exist, but I would like to tackle them separately:

- Look at enabling the flags I've disabled in this PR, most notably `--authentication-config`, which seems like a hugely useful feature for a multi-tenant API server like KCP.

## Related issue(s)

Fixes #3068

## Release Notes

```release-note
Update to Kubernetes 1.30.0
```

### CODE REVIEW

**Comment by mjudeikis** [comment]
🥳 🔥 🎆

**Comment by mjudeikis** [comment]
/lgtm
/approve
/hold
@sttts wanna read through or fix on master? :)

File: pkg/admission/apiresourceschema/validation_test.go [hunk 1]
```diff
@@ -23,12 +23,14 @@ import (
...
-		"PreexistingExpressions": reflect.Struct,
+		"DisallowDefaultsReason":       reflect.String,
+		"CELEnvironmentSet":            reflect.Ptr,
+		"PreexistingExpressions":       reflect.Struct,
+		"VersionsWithUnchangedSchemas": reflect.Map,
 	}
 	expectedFalse := map[string]bool{
-		"RequireImmutableNames": true,
+		"RequireImmutableNames":     true,
+		"SuppressPerExpressionCost": true,
```
Comment (sttts) [inline pkg/admission/apiresourceschema/validation_test.go:33]: remind me why this change is correct?

File: pkg/admission/plugins.go [hunk 2]
```diff
@@ -174,7 +174,7 @@ var defaultOnKubePluginsInKube = sets.New[string](
 	defaulttolerationseconds.PluginName,     // DefaultTolerationSeconds
 	mutatingwebhook.PluginName,              // MutatingAdmissionWebhook
 	validatingwebhook.PluginName,            // ValidatingAdmissionWebhook
-	validatingadmissionpolicy.PluginName,    // ValidatingAdmissionPolicy
+	validating.PluginName,                   // ValidatingAdmissionPolicy
```
Comment (sttts) [inline pkg/admission/plugins.go:177]: the old import alias was better

File: pkg/cache/server/config.go [hunk 3]
```diff
@@ -148,7 +146,6 @@ func NewConfig(opts *cacheserveroptions.CompletedOptions, optionalLocalShardRest
 		return apiHandler
 	}
 
-	opts.Etcd.StorageConfig.Paging = utilfeature.DefaultFeatureGate.Enabled(features.APIListChunking)
```
Comment (sttts) [inline pkg/cache/server/config.go:151]: is this field gone or should it be true by default?

File: pkg/proxy/options/authentication.go [hunk 4]
```diff
@@ -121,7 +122,7 @@ func (c *Authentication) ApplyTo(authenticationInfo *genericapiserver.Authentica
 	}
 
 	// Sets up a union Authenticator for all enabled auth methods
-	authenticationInfo.Authenticator, _, err = authenticatorConfig.New()
+	authenticationInfo.Authenticator, _, _, _, err = authenticatorConfig.New(context.TODO()) // TODO(embik): wire server context in?
```
Comment (sttts) [inline pkg/proxy/options/authentication.go:125]: needs proper context, yes

File: pkg/server/openapiv3/controller.go [hunk 5]
```diff
@@ -192,22 +193,26 @@ func (c *Controller) processCRD(crd *apiextensionsv1.CustomResourceDefinition) {
...
+				}
+				bs, err := json.Marshal(spec)
+				if err != nil {
+					return nil, "", err
+				}
+				return spec, fmt.Sprintf("%X", sha512.Sum512(bs)), nil
+			},
+		}
+
+		spec := cached.Once[*spec3.OpenAPI](val)
```
Comment (sttts) [inline pkg/server/openapiv3/controller.go:210]:
```suggestion
		spec := cached.Once(cached.Func[*spec3.OpenAPI](
			func() (value *spec3.OpenAPI, etag string, err error) {
				spec, err := builder.BuildOpenAPIV3(crd, v.Name, builder.Options{V2: false})
				if err != nil {
					return nil, "", err
				}
				bs, err := json.Marshal(spec)
				if err != nil {
					return nil, "", err
				}
				return spec, fmt.Sprintf("%X", sha512.Sum512(bs)), nil
			},
		))
```

File: pkg/openapi/value.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+*/
+
+package openapi
+
+import (
+	"k8s.io/kube-openapi/pkg/cached"
+	"k8s.io/kube-openapi/pkg/spec3"
+)
+
+type StaticValue[T any] struct {
```
Comment (sttts) [inline pkg/openapi/value.go:24]: no need for this, see my comment where the spec is computed.

File: pkg/reconciler/topology/partitionset/resources.go [hunk 7]
```diff
@@ -51,11 +51,11 @@ func generatePartitionName(name string, matchLabels map[string]string, dimension
 	for _, label := range labels {
 		name = name + "-" + strings.ToLower(matchLabels[label])
 	}
-	name = name[:min(validation.DNS1123SubdomainMaxLength-1, len(name))]
+	name = name[:minimum(validation.DNS1123SubdomainMaxLength-1, len(name))]
 	return name
 }
 
-func min(a, b int) int {
+func minimum(a, b int) int {
```
Comment (sttts) [inline pkg/reconciler/topology/partitionset/resources.go:58]: can we just remove our `min`? It's not in the language.

File: pkg/server/openapiv3/servicecache.go [hunk 8]
```diff
@@ -101,15 +103,20 @@ func (c *ServiceCache) RegisterStaticAPIs(cont *restful.Container) error {
...
+				if err != nil {
+					return nil, "", fmt.Errorf("failed to build OpenAPI v3 spec for %s: %w", gvPath, err)
+				}
+				etag, err = computeEtag(spec)
+				if err != nil {
+					return nil, "", fmt.Errorf("failed to compute OpenAPI v3 spec etag for %s: %w", gvPath, err)
+				}
+
+				return spec, etag, nil
+			},
```
Comment (sttts) [inline pkg/server/openapiv3/servicecache.go:118]: this doesn't look correct. We should only call `GetFunc` once, i.e. probably this will fit here too:
```go
cached.Once(cached.Func[*spec3.OpenAPI](
			func() (value *spec3.OpenAPI, etag string, err error) {
...
```

File: pkg/server/options/flags.go [hunk 9]
```diff
@@ -192,5 +193,13 @@ var (
...
 		"peer-ca-file",                          // If set and the UnknownVersionInteroperabilityProxy feature gate is enabled, this file will be used to verify serving certificates of peer kube-apiservers. This flag is only used in clusters configured with multiple kube-apiservers for high availability.
+
+		// logs flags
+		"log-text-info-buffer-size", // [Alpha] In text format with split output streams, the info messages can be buffered for a while to increase performance. The default value of zero bytes disables buffering. The size can be specified as number of bytes (512), multiples of 1000 (1K), multiples of 1024 (2Ki), or powers of those (3M, 4G, 5Mi, 6Gi). Enable the LoggingAlphaOptions feature gate to use this.
+		"log-text-split-stream",     // [Alpha] In text format, write error messages to stderr and info messages to stdout. The default is to write a single stream to stdout. Enable the LoggingAlphaOptions feature gate to use this.
+
+		// authentication flags
+		// TODO(embik): look at enabling this feature.
```
Comment (sttts) [inline pkg/server/options/flags.go:202]: as this will likely not work out of the box, should we move it out of allowed flags?

File: pkg/virtual/apiexport/schemas/builtin/builtin.go [hunk 10]
```diff
@@ -261,6 +272,16 @@ var BuiltInAPIs = []internalapis.InternalAPI{
 		Instance:      &admissionregistrationv1alpha1.ValidatingAdmissionPolicyBinding{},
 		ResourceScope: apiextensionsv1.ClusterScoped,
 	},
+	{
+		Names: apiextensionsv1.CustomResourceDefinitionNames{
+			Plural:   "validatingadmissionpolicybindings",
```
[the remark this answers is not in the document]
Comment (embik) [reply pkg/virtual/apiexport/schemas/builtin/builtin.go:277]: I definitely want an e2e test for this (there's a note on it in the PR description), but I didn't want to squeeze it into this PR. Do you think it would be better to do this immediately?

File: pkg/server/options/flags.go (hunk 9 above)
Comment (embik) [reply pkg/server/options/flags.go:202]: This is the `disallowedFlags` set :D

File: pkg/cache/server/config.go (hunk 3 above)
Comment (embik) [reply pkg/cache/server/config.go:151]: The option is completely gone.

Comment (embik) [reply pkg/cache/server/config.go:151]: Pagination is enforced since https://github.com/kubernetes/kubernetes/pull/121390 if I understand correctly.

File: pkg/reconciler/topology/partitionset/resources.go (hunk 7 above)
Comment (embik) [reply pkg/reconciler/topology/partitionset/resources.go:58]: Good point. I forgot that there is the builtin `min` now ...

File: pkg/server/options/flags.go (hunk 9 above)
Comment (sttts) [reply pkg/server/options/flags.go:202]: ah 😀

File: pkg/virtual/apiexport/schemas/builtin/builtin.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/virtual/apiexport/schemas/builtin/builtin.go:277]: If the API is enabled, we should test it. We can disable it and postpone to followup.

File: pkg/admission/apiresourceschema/validation_test.go (hunk 1 above)
Comment (embik) [reply pkg/admission/apiresourceschema/validation_test.go:33]: I'm not 100% sure here, appreciate more input, but since it's a new field I thought that "expecting it to be false" was a reasonable assumption. This is the field in question:

```go
// versionsWithUnchangedSchemas tracks schemas of which versions are unchanged when updating a CRD.
// Does not apply to creation or deletion.
// Some checks use this to avoid rejecting previously accepted versions due to a control plane upgrade/downgrade.
versionsWithUnchangedSchemas sets.Set[string]
// suppressPerExpressionCost indicates whether CEL per-expression cost limit should be suppressed.
// It will be automatically set during Versions validation if the version is in versionsWithUnchangedSchemas.
suppressPerExpressionCost bool
```

I didn't see it as important that it's set.

Comment (sttts) [reply pkg/admission/apiresourceschema/validation_test.go:33]: The test here is to force us understand which values new fields should have.

False is correct. Please add the reasoning to `defaultValidationOpts`:
```golang
	// in Kube, the validation on CRD update will set it to true for unchanged schemas
	// (ratcheting validation). Here, as everything is immutable, this case never
	// happens. Hence, we can statically set it to false.
	SuppressPerExpressionCost: false,
```

File: pkg/virtual/apiexport/schemas/builtin/builtin.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (embik) [reply pkg/virtual/apiexport/schemas/builtin/builtin.go:277]: Test has been added. It was in fact not working, I had to fix things on k8s and kcp side. Key commits:

- https://github.com/kcp-dev/kubernetes/commit/d61cdaca2a4ad0b8e57f31325f3aaf4c507bfec6
- https://github.com/kcp-dev/kubernetes/commit/3b450c7d0daf983858646073b08aca3e0ed9ba9e 
- https://github.com/kcp-dev/kcp/pull/3140/commits/bcf32edaa5456d0bc9e5208534a17313e3793154
- https://github.com/kcp-dev/kcp/pull/3140/commits/6af5c9314b07d08b0962ba11599779c8df179e49

File: test/e2e/conformance/validatingadmissionpolicy_test.go [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+	orgPath, _ := framework.NewOrganizationFixture(t, server)
+	ws1Path, _ := framework.NewWorkspaceFixture(t, server, orgPath)
+	ws2Path, _ := framework.NewWorkspaceFixture(t, server, orgPath)
+	paths := []logicalcluster.Path{ws1Path, ws2Path}
+
+	kubeClusterClient, err := kcpkubernetesclientset.NewForConfig(cfg)
+	require.NoError(t, err, "failed to construct client for server")
+	cowbyClusterClient, err := wildwestclientset.NewForConfig(cfg)
+	require.NoError(t, err, "failed to construct cowboy client for server")
+	apiExtensionsClients, err := kcpapiextensionsclientset.NewForConfig(cfg)
```
Comment (sttts) [inline test/e2e/conformance/validatingadmissionpolicy_test.go:71]: cluster client?

File: test/e2e/conformance/validatingadmissionpolicy_test.go [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+	require.NoError(t, err, "failed to add admission registration v1 scheme")
+	err = v1.AddToScheme(scheme)
+	require.NoError(t, err, "failed to add admission v1 scheme")
+	err = wildwestv1alpha1.AddToScheme(scheme)
+	require.NoError(t, err, "failed to add cowboy v1alpha1 to scheme")
+
+	orgPath, _ := framework.NewOrganizationFixture(t, server)
+	ws1Path, _ := framework.NewWorkspaceFixture(t, server, orgPath)
+	ws2Path, _ := framework.NewWorkspaceFixture(t, server, orgPath)
+	paths := []logicalcluster.Path{ws1Path, ws2Path}
```
Comment (sttts) [inline test/e2e/conformance/validatingadmissionpolicy_test.go:65]: am not a fan of these indirections. Just write this out where the list is needed and reference `ws1/2Path` directly.

File: test/e2e/conformance/validatingadmissionpolicy_test.go [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+				},
+			},
+			Validations: []admissionregistrationv1.Validation{{
+				Expression: "object.spec.intent != 'bad'",
+			}},
+		},
+	}
+	policy, err = kubeClusterClient.Cluster(paths[0]).AdmissionregistrationV1().ValidatingAdmissionPolicies().Create(ctx, policy, metav1.CreateOptions{})
+
+	require.NoError(t, err, "failed to create ValidatingAdmissionPolicy")
```
Comment (sttts) [inline test/e2e/conformance/validatingadmissionpolicy_test.go:112]:
```suggestion
	policy, err = kubeClusterClient.Cluster(paths[0]).AdmissionregistrationV1().ValidatingAdmissionPolicies().Create(ctx, policy, metav1.CreateOptions{})
	require.NoError(t, err, "failed to create ValidatingAdmissionPolicy")
```

File: test/e2e/conformance/validatingadmissionpolicy_test.go [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+		},
+	}
+
+	_, err = kubeClusterClient.Cluster(paths[0]).AdmissionregistrationV1().ValidatingAdmissionPolicyBindings().Create(ctx, binding, metav1.CreateOptions{})
+	require.NoError(t, err, "failed to create ValidatingAdmissionPolicyBinding")
+
+	// Policies take a second to compile. So far there doesn't seem to be a way
+	// to query if a policy has been successfully processed and is active, so we
+	// just sleep for a few seconds.
+	time.Sleep(5 * time.Second)
```
Comment (sttts) [inline test/e2e/conformance/validatingadmissionpolicy_test.go:142]: timeouts always break eventually. Better just repeat the check below until the error appears.

File: test/e2e/conformance/validatingadmissionpolicy_test.go [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "cowboy",
+		},
+		Spec: wildwestv1alpha1.CowboySpec{
+			Intent: "good",
+		},
+	}
+
+	// badCowboy violates the policy, so it should be rejected
+	t.Logf("Creating bad cowboy resource in first logical cluster")
```
Comment (sttts) [inline test/e2e/conformance/validatingadmissionpolicy_test.go:163]: style: combine both Log and comment. Having both is noisy.

**Comment by sttts** [comment]
Modulo the e2e style comments, lgtm.

File: test/e2e/conformance/validatingadmissionpolicy_test.go (hunk 12 above)
Comment (embik) [reply test/e2e/conformance/validatingadmissionpolicy_test.go:65]: Copied this part from another of the conformance tests, but I agree. Will adjust it here.

**Comment by embik** [comment]
Looks like the failing tests aren't a flake:

```
maximalpermissionpolicy_authorizer_test.go:232: 
        	Error Trace:	/home/prow/go/src/github.com/kcp-dev/kcp/test/e2e/apibinding/maximalpermissionpolicy_authorizer_test.go:404
        	            				/home/prow/go/src/github.com/kcp-dev/kcp/test/e2e/apibinding/maximalpermissionpolicy_authorizer_test.go:232
        	Error:      	Received unexpected error:
        	            	cowboys.wildwest.dev "cowboy-e2e-workspace-fhj5z" is forbidden: ValidatingAdmissionPolicy 'policy-4lr9q' with binding 'binding-4vxvf' denied request: expression 'object.spec.intent != 'bad'' resulted in error: no such key: intent
        	Test:       	TestMaximalPermissionPolicyAuthorizer
        	Messages:   	error creating cowboy in consumer workspace "root:e2e-workspace-jmr78:e2e-workspace-fhj5z"
```

This isn't always triggered, but it needs to be addressed.

File: pkg/admission/validatingadmissionpolicy/validating_admission_policy.go [hunk 16]
```diff
@@ -136,7 +141,8 @@ func (k *KubeValidatingAdmissionPolicy) Validate(ctx context.Context, a admissio
 		return err
 	}
 
-	return delegate.Validate(ctx, a, o)
+	err = delegate.Validate(ctx, a, o)
+	return err
```
Comment (sttts) [inline pkg/admission/validatingadmissionpolicy/validating_admission_policy.go:145]: nit: the previous was fine

File: pkg/admission/validatingadmissionpolicy/validating_admission_policy.go [hunk 17]
```diff
@@ -189,6 +193,20 @@ func (k *KubeValidatingAdmissionPolicy) getOrCreateDelegate(clusterName logicalc
 	plugin.SetDrainedNotification(ctx.Done())
 	plugin.InspectFeatureGates(k.featureGates)
 	plugin.SetAuthorizer(k.authorizer)
+	plugin.SetClusterName(clusterName)
+	plugin.SetSourceFactory(func(_ informers.SharedInformerFactory, client kubernetes.Interface, dynamicClient dynamic.Interface, restMapper meta.RESTMapper, clusterName logicalcluster.Name) generic.Source[validating.PolicyHook] {
+		return generic.NewPolicySource(
+			k.globalKubeSharedInformerFactory.Admissionregistration().V1().ValidatingAdmissionPolicies().Informer().Cluster(clusterName),
+			k.globalKubeSharedInformerFactory.Admissionregistration().V1().ValidatingAdmissionPolicyBindings().Informer().Cluster(clusterName),
```
Comment (sttts) [inline pkg/admission/validatingadmissionpolicy/validating_admission_policy.go:200]: Did you add replication of policies? I would have expected this to fail if not. The local informers have the local shard objects.

File: test/e2e/conformance/validatingadmissionpolicy_test.go [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+					},
+				},
+			},
+			Validations: []admissionregistrationv1.Validation{{
+				Expression: "object.spec.intent != 'bad'",
+			}},
+		},
+	}
+	policy, err = kubeClusterClient.Cluster(ws1Path).AdmissionregistrationV1().ValidatingAdmissionPolicies().Create(ctx, policy, metav1.CreateOptions{})
+
```
Comment (sttts) [inline test/e2e/conformance/validatingadmissionpolicy_test.go:112]: nit/style: no empty line before error checks.

File: test/e2e/conformance/validatingadmissionpolicy_test.go [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+	t.Logf("Verifying that creating bad cowboy resource in first logical cluster is rejected")
+	require.Eventually(t, func() bool {
+		_, err := cowbyClusterClient.Cluster(ws1Path).WildwestV1alpha1().Cowboys("default").Create(ctx, &badCowboy, metav1.CreateOptions{})
+		if err != nil {
+			if errors.IsInvalid(err) {
+				if strings.Contains(err.Error(), "failed expression: object.spec.intent != 'bad'") {
+					return true
+				}
+			}
+			t.Logf("Unexpected error when trying to create bad cowboy: %s", err)
```
Comment (sttts) [inline test/e2e/conformance/validatingadmissionpolicy_test.go:167]: nit: a success is not unexpected while waiting for the policy to apply.

Comment (embik) [reply test/e2e/conformance/validatingadmissionpolicy_test.go:167]: This is within a check for `err != nil`, so it only logs any error that's not the policy rejection. If resource creation succeeds, it just returns `false` but doesn't log this line.

File: pkg/admission/validatingadmissionpolicy/validating_admission_policy.go (hunk 17 above)
Comment (embik) [reply pkg/admission/validatingadmissionpolicy/validating_admission_policy.go:200]: https://github.com/kcp-dev/kcp/pull/3140/files#diff-56ef0431fd7b0e03ab56df4aba923c02c16a6e523bd27230cd3aa856203c5f9eR228

This is where replication is configured. Through `InstallIndexers`, this is is wired into `(s *Server) addIndexersToInformers`, which returns the list of replicated GVRs. This is a relatively recent change made in #3139, which was necessary to that informer installation doesn't create a race condition on leader election changes.

**Comment by embik** [comment]
/retest

flake.

File: pkg/admission/validatingadmissionpolicy/validating_admission_policy.go (hunk 17 above)
Comment (sttts) [reply pkg/admission/validatingadmissionpolicy/validating_admission_policy.go:199]: ack
